ttg/ttg/serialization/backends/cereal.h

```diff
@@ -0,0 +1,98 @@
+
+  template <typename Archive, typename T>
+  inline static constexpr bool is_cereal_serializable_v = is_cereal_serializable<Archive, T>::value;
+
+  template <typename T, class = void>
+  struct is_cereal_buffer_serializable : std::false_type {};
+
+#ifdef TTG_SERIALIZATION_SUPPORTS_CEREAL
+  template <typename T>
+  struct is_cereal_buffer_serializable<T, std::enable_if_t<is_cereal_serializable_v<cereal::BinaryInputArchive, T> &&
+                                                           is_cereal_serializable_v<cereal::BinaryOutputArchive, T>>>
+      : std::true_type {};
+#endif  // TTG_SERIALIZATION_SUPPORTS_CEREAL
+
+  /// evaluates to true if can serialize @p T to/from buffer using Cereal serialization
+  template <typename T>
+  inline constexpr bool is_cereal_buffer_serializable_v = is_cereal_buffer_serializable<T>::value;
+
+  template <typename Archive, typename T, typename Enabler = void>
+  struct is_cereal_array_serializable;
+
+  template <typename Archive, typename T>
+  struct is_cereal_array_serializable<Archive, T, std::enable_if_t<!std::is_array<T>::value>> : std::false_type {};
+
+  template <typename Archive, typename T>
+  struct is_cereal_array_serializable<Archive, T, std::enable_if_t<std::is_array<T>::value>>
+      : std::bool_constant<is_cereal_serializable_v<Archive, std::remove_extent_t<T>>> {};
+
+  template <typename Archive, typename T>
+  inline static constexpr bool is_cereal_array_serializable_v = is_cereal_array_serializable<Archive, T>::value;
+
+  template <typename Archive, typename T>
+  inline static constexpr bool is_stlcontainer_cereal_serializable_v = false;
+
+  template <typename Archive, typename T, class = void>
+  struct is_cereal_user_serializable : std::false_type {};
+
+#ifdef TTG_SERIALIZATION_SUPPORTS_CEREAL
+  template <typename Archive, typename T>
+  struct is_cereal_user_serializable<
+      Archive, T,
+      std::enable_if_t<(cereal::traits::detail::count_specializations<T, Archive>::value != 0) ||
+                       ((cereal::traits::is_input_serializable<T, Archive>::value ||
+                         cereal::traits::is_output_serializable<T, Archive>::value) &&
+                        (!std::is_arithmetic_v<T> && !ttg::detail::is_cereal_array_serializable_v<Archive, T> &&
+                         !is_stlcontainer_cereal_serializable_v<Archive, T>))>> : std::true_type {};
+#endif
+
+  template <typename Archive, typename T>
+  inline constexpr bool is_cereal_user_serializable_v = is_cereal_user_serializable<Archive, T>::value;
+
+  template <typename T, class = void>
+  struct is_cereal_user_buffer_serializable : std::false_type {};
+
+#ifdef TTG_SERIALIZATION_SUPPORTS_CEREAL
+  template <typename T>
+  struct is_cereal_user_buffer_serializable<
+      T, std::enable_if_t<is_cereal_user_serializable_v<cereal::BinaryInputArchive, T> ||
+                          is_cereal_user_serializable_v<cereal::BinaryOutputArchive, T>>> : std::true_type {};
```

Comment:
Shouldn't we test with `&&` here? We need support for both input and output archive from T, no?

Reply:
If user provided serialization method for either we should insist on using user serializability. If only one was provided this will likely lead to compile-time error, or the user indeed wanted this to happen